windows-installer/auto-update.ps1

```diff
@@ -0,0 +1,134 @@
+[CmdletBinding()]
+param()
+begin {}
+process {
+    New-EventLog -Source "certbot/auto-update.ps1" -LogName "CertbotAutoUpdate" -ErrorAction SilentlyContinue
+
+    function Write-Message($message, $level = "Information") {
+        Write-EventLog -Source "certbot/auto-update.ps1" -LogName "CertbotAutoUpdate" -EventID 1 -EntryType $level -Message $message
+        Write-Host $message
+    }
+
+    function Write-Error($message) {
+        Write-EventLog -Source "certbot/auto-update.ps1" -LogName "CertbotAutoUpdate" -EventID 1 -EntryType Error -Message $message
+        throw $message
+    }
+
+    Write-Message "Starting auto-update workflow ..."
+
+    $ErrorActionPreference = 'Stop'
+
+    $installDir = $PSScriptRoot
+
+    if ((Test-Path HKLM:\Software\Certbot) -And ((Get-ItemProperty -Path HKLM:\Software\Certbot).PSObject.Properties.Name -Contains "CertbotSigningPubKey")) {
+        $certbotSigningPubKey = (Get-ItemProperty -Path HKLM:\Software\Certbot).CertbotSigningPubKey
+    } else {
+        $certbotSigningPubKey = '
+-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA6MR8W/galdxnpGqBsYbq
+OzQb2eyW15YFjDDEMI0ZOzt8f504obNs920lDnpPD2/KqgsfjOgw2K7xWDJIj/18
+xUvWPk3LDkrnokNiRkA3KOx3W6fHycKL+zID7zy+xZYBuh2fLyQtWV1VGQ45iNRp
+9+Zo7rH86cdfgkdnWTlNSHyTLW9NbXvyv/E12bppPcEvgCTAQXgnDVJ0/sqmeiij
+n9tTFh03aM+R2V/21h8aTraAS24qiPCz6gkmYGC8yr6mglcnNoYbsLNYZ69zF1XH
+cXPduCPdPdfLlzVlKK1/U7hkA28eG3BIAMh6uJYBRJTpiGgaGdPd7YekUB8S6cy+
+CQIDAQAB
+-----END PUBLIC KEY-----
+'
+    }
+
+    if ((Test-Path HKLM:\Software\Certbot) -And ((Get-ItemProperty -Path HKLM:\Software\Certbot).PSObject.Properties.Name -Contains "CertbotUpgradeApiURL")) {
+        $certbotUpgradeApiURL = (Get-ItemProperty -Path HKLM:\Software\Certbot).CertbotUpgradeApiURL
+    } else {
+        $certbotUpgradeApiURL = 'https://api.github.com/repos/certbot/certbot/releases/latest'
+    }
+
+    # Get current local certbot version
+    try {
+        $currentVersion = certbot --version
+        $currentVersion = $currentVersion -replace '^certbot (\d+\.\d+\.\d+).*$', '$1'
+        $currentVersion = [System.Version]"$currentVersion"
+    } catch {
+        Write-Message @"
+An error occured while fetching the current local certbot version:
+$_
+Assuming Certbot is not up-to-date.
+"@ "Warning"
+        $currentVersion = "0.0.0"
+    }
+
+    # Get latest remote certbot version
+    try {
+        [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
```

Comment:
Initially I added this line because of Windows 2012 R2: GitHub accepts only TLS 1.1+ and only TLS 1.0 was enabled for this system. At the basic level, then this line is not needed anymore because we do not support this system.

Now I think there is still something to do. At least we need to have the line `[Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12)`, because it will enable the use of TLS 1.2 at least on all systems, or a better TLS version if it is available.

Now if GitHub decides on day to accept only TLS 1.3, this code will not break either if:
* the systems have all TLS 1.3 available and use it by default
* or we update the current script to enable TLS 1.3 or better, with `[Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls13)`

However we cannot do anything for systems that do not have TLS 1.3 available.

I replace the line by `[Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12)`.

Reply:
I looked into this a bit more and think we may still have a problem. According to https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.securityprotocol?view=netframework-4.8, since .NET 4.7, the default value of `[Net.ServicePointManager]::SecurityProtocol` is [SecurityProtocolType.SystemDefault](https://docs.microsoft.com/en-us/dotnet/api/system.net.securityprotocoltype?view=netframework-4.8#System_Net_SecurityProtocolType_SystemDefault) whose numerical value is 0. See:
```
PS /> [Net.ServicePointManager]::SecurityProtocol
SystemDefault
PS /> [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12
Tls12
```
This to me makes it look like it is not possible to add TLS 1.2 to the list of protocols the OS supports by default and trying to do so makes Powershell only support that protocol.

Unless I'm misunderstanding how this works or there is a better way to add TLS 1.2 to the list of protocols, I'm personally towards leaving this with its default value. What do you think?

Reply:
Too bad. It was not working like this on my machine, and the value was an explicit composition of tls values. Well, better to remove that then, it works anyway out of the box for now with the supported Windows versions.